lib/using_restricted_array.rb

```diff
@@ -1,61 +1,114 @@
+# Time complexity: O(n) where n is the number of values in the array
+# Space complexity: O(1), constant regardless of size of input
 def print_array(array)
-  raise NotImplementedError
+  i = 0
+  while i < len
+    print array[i]
+    print ""
+    i += 1
+  end
+  puts
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the number of values in the array
+# Space complexity: O(1), constant regardless of size of input
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  # raise NotImplementedError
+  length.times do |i|
+    return true if array[i] == value_to_find
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the number of values in the array
+# Space complexity: O(1), constant regardless of size of input
 def find_largest(array, length)
-  raise NotImplementedError
+  largest = array[0]
+  index = 0
+  while index < length
+    if largest < array[index]
+      largest = array[index]
+    end
+    index += 1
+  end
+  return largest
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the number of values in the array
+# Space complexity: O(1), constant regardless of size of input
 def find_smallest(array, length)
-  raise NotImplementedError
+  smallest = array[0]
+  index = 0
+  while index < length
+    if smallest > array[index]
+      smallest = array[index]
+    end
+    index += 1
+  end
+  return smallest
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the number of values in the array
+# Space complexity: O(n), where n is the size of the input
 def reverse(array, length)
-  raise NotImplementedError
+  smaller_index = 0
+  larger_index = length - 1
+
+  until smaller_index > larger_index
+    temp = array[smaller_index]
+    array[smaller_index] = array[larger_index]
+    array[larger_index] = temp
+    smaller_index += 1
+    larger_index -= 1
+  end
+  return array
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(log n) where n is the number of values in the array
+# Space complexity: O(n), where n is the size of the input
 def binary_search(array, length, value_to_find)
-  raise NotImplementedError
+  smaller = 0
+  larger = length - 1
+
+  until smaller > larger
+    mid = (smaller + larger) / 2
+    if array[mid] == value_to_find
+      return true
+    elsif array[mid] < value_to_find
+      smaller = mid
+    elsif array[mid] > value_to_find
```

Comment:
This can just be `else` since you're already checking `==` and `<` in line 102 and 105. It is a slight optimization that saves an extra variable comparison.